src/dim_helpers/ConvDims.jl

```diff
@@ -6,34 +6,54 @@ export ConvDims
 Type system-level information about convolution dimensions. Critical for things like
 `im2col!()` to generate efficient code, and helpful to reduce the number of kwargs
 getting passed around.
-
-We don't want to specialize on things like image size/channel count, so we generally
-store those as fields, just for convenience, and to allow for non-breaking changes when
-we decide we _do_ want to specialize on those values.  We always want to specialize on
-things like stride, padding, dilation, and kernel flipping though.
 """
-abstract type ConvDims{N, S, P, D, F} end
+abstract type ConvDims{N} end
+
+@inline spatial_dims(::ConvDims{N}) where N = N
+@inline groupcount(c::ConvDims) = 1
+
+function input_size(::ConvDims) end
+function kernel_size(::ConvDims) end
+
+function stride(::ConvDims) end
+function padding(::ConvDims) end
+function dilation(::ConvDims) end
+function flipkernel(::ConvDims) end
 
-# Hack to get rid of type parameters
 function basetype(::Type{C}) where {C <: ConvDims}
-    if C <: DepthwiseConvDims
-        return DepthwiseConvDims
-    elseif C <: DenseConvDims
+    if C <: DenseConvDims
```

Comment:
is it necessary to change this order?

Reply:
No, this was due to some intermediate experimentation. Reverted.